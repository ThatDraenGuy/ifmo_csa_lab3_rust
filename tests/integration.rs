use std::{
    error::Error,
    fs::{self, File},
    io::Read,
    path::Path,
};

use gag::BufferRedirect;
use insta::{assert_ron_snapshot, glob};
use lab3_rust::{machine, translator};
use serde::{Deserialize, Serialize};
use tempfile::{tempdir, TempDir};

#[derive(Deserialize)]
struct TestInput {
    translator_input: String,
    machine_input: String,
}

#[derive(Serialize)]
struct TestOutput {
    translator_output: String,
    stdout: String,
}

fn perform_test(temp_dir: &TempDir, input_path: &Path) -> Result<TestOutput, Box<dyn Error>> {
    // Достаём имя теста
    let test_name = input_path.file_stem().unwrap().to_str().unwrap();

    // Читаем входные данные
    let input_file = File::open(input_path)?;
    let input: TestInput = serde_yaml::from_reader(input_file)?;

    // Создаём временные файлы для входных и выходных данных
    let translator_input = temp_dir.path().join(test_name.to_owned() + "__translator_in");
    let machine_input = temp_dir.path().join(test_name.to_owned() + "__machine_in");
    let target = temp_dir.path().join(test_name.to_owned() + "__target");

    // Копируем входные данные
    fs::write(&translator_input, input.translator_input)?;
    fs::write(&machine_input, input.machine_input)?;

    // Хватаем вывод в STDOUT
    let mut stdout_buf = BufferRedirect::stdout()?;
    // Запускаем транслятор и виртуальную машину
    translator::main(&translator_input, &target)?;
    machine::main(&target, &machine_input)?;

    // Читаем выходные данные
    let mut stdout = String::new();
    stdout_buf.read_to_string(&mut stdout)?;

    // Читаем результат работы транслятора
    let translator_output = fs::read_to_string(target)?;

    //Возвращаем результат
    Ok(TestOutput { translator_output, stdout })
}

#[test]
fn test() {
    // Создаём одну временную директорию для всех файлов
    let dir = tempdir().unwrap();
    glob!("inputs/*.ron", |input_path| {
        // Проходимся по всем файлам с входными данными и для каждого из них создаём снапшот
        let result = perform_test(&dir, input_path).unwrap();
        assert_ron_snapshot!(result);
    })
}
