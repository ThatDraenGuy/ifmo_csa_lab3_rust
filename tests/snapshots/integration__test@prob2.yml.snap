---
source: tests/integration.rs
expression: result
input_file: tests/inputs/prob2.yml
---
translator_output: |-
  (
      code: [
          Data((1)),
          Data((2)),
          OpHigher(Math((
              opcode: Mov,
              args: RegImmed(Count, 0),
          ))),
          OpLower(MathOpLower(RegImmed(4))),
          OpHigher(Math((
              opcode: Mov,
              args: RegImmed(Accumulator, 0),
          ))),
          OpLower(MathOpLower(RegImmed(2))),
          OpHigher(Math((
              opcode: Mov,
              args: RegToReg(Data, Accumulator),
          ))),
          OpHigher(Math((
              opcode: Add,
              args: RegToMem(Accumulator, 0),
          ))),
          OpLower(MathOpLower(RegToMem(1))),
          OpHigher(Math((
              opcode: Mov,
              args: MemToReg(Data, 0),
          ))),
          OpLower(MathOpLower(MemToReg(1))),
          OpHigher(Alter((
              opcode: Dec,
              arg: Count,
          ))),
          OpHigher(Branch((
              opcode: Jnz,
              arg_higher: 0,
          ))),
          OpLower(BranchOpLower((
              arg_lower: 7,
          ))),
          OpHigher(Math((
              opcode: Mov,
              args: RegImmed(Count, 0),
          ))),
          OpLower(MathOpLower(RegImmed(4))),
          OpHigher(Math((
              opcode: Cmp,
              args: RegImmed(Accumulator, 61),
          ))),
          OpLower(MathOpLower(RegImmed(2304))),
          OpHigher(Branch((
              opcode: Jns,
              arg_higher: 0,
          ))),
          OpLower(BranchOpLower((
              arg_lower: 25,
          ))),
          OpHigher(Math((
              opcode: Add,
              args: MemToReg(Accumulator, 0),
          ))),
          OpLower(MathOpLower(MemToReg(2))),
          OpHigher(Branch((
              opcode: Jmp,
              arg_higher: 0,
          ))),
          OpLower(BranchOpLower((
              arg_lower: 7,
          ))),
          OpHigher(Math((
              opcode: Mov,
              args: RegToMem(Accumulator, 0),
          ))),
          OpLower(MathOpLower(RegToMem(2))),
          OpHigher(Io((
              opcode: Out,
              arg: 1,
          ))),
          OpHigher(Control((Exit))),
      ],
      entrypoint: (3),
  )
stdout: ''

